experiments/toptagging/wrappers.py

```diff
@@ -31,6 +32,199 @@ def attention_mask(batch, force_xformers=True):
         mask = mask.materialize(shape=(len(batch.batch), len(batch.batch)))
     return mask
 
+def attention_mask_pretrain(attention_mask_list, force_xformers=True):
+    """
+    Construct attention mask that makes sure that objects only attend to each other
+    within the same batch element, and not across batch elements
+
+    Parameters
+    ----------
+    batch: torch_geometric.data.Data
+        torch_geometric representation of the data
+    force_xformers: bool
+        Decides whether a xformers or torch.tensor mask should be returned
+        The xformers mask allows to use the optimized xformers attention kernel, but only runs on gpu
+
+    Returns
+    -------
+    mask: xformers.ops.fmha.attn_bias.BlockDiagonalMask or torch.tensor
+        attention mask, to be used in xformers.ops.memory_efficient_attention
+    """
+    bincounts = torch.bincount(attention_mask_list).tolist()
+    mask = BlockDiagonalMask.from_seqlens(bincounts)
+    if not force_xformers:
+        # materialize mask to torch.tensor (only for testing purposes)
+        mask = mask.materialize(shape=(len(attention_mask_list), len(attention_mask_list)))
+    return mask
+
+
+class TopTaggingTransformerWrapper(nn.Module):
+    """
+    Baseline Transformer for top-tagging
+    This is mainly for debugging
+    """
+
+    def __init__(self, net, force_xformers=True):
+        super().__init__()
+        self.net = net
+        self.force_xformers = force_xformers
+
+    def forward(self, batch):
+        mask = attention_mask(batch, self.force_xformers)
+
+        # unsqueeze and squeeze to formally add the batch index (any better way of doing this?)
+        inputs = batch.x.unsqueeze(0)
+        outputs = self.net(inputs, attention_mask=mask)
+        logits = outputs.squeeze(0)[batch.is_global]
+
+        return logits
+
+class TopTaggingPretrainGATrWrapper(nn.Module):
+    """
+    GATr for toptagging
+    including all kinds of options to play with
+    """
+
+    def __init__(
+        self,
+        net,
+        mean_aggregation=False,
+        add_pt=False,
+        force_xformers=True,
+        add_jet_momentum=False,
+        beam_reference="spacelike",
+        add_time_reference=False,
+        two_beams=False,
+        beam_token=True,
+    ):
+        super().__init__()
+        self.net = net
+        self.mean_aggregation = mean_aggregation
+        self.add_pt = add_pt
+        self.force_xformers = force_xformers
+        self.add_jet_momentum = add_jet_momentum
+        self.beam_reference = beam_reference
+        self.add_time_reference = add_time_reference
+        self.two_beams = two_beams
+        self.beam_token = beam_token
+
+    def forward(self, batch):
```

Comment:
Instead of having this big method with the wild (and probably very expensive) for loop, we could do this in the dataset class, similar to the TopTaggingDataset and QGTaggingDataset. (As a side note, we could unify these two a bit because they share a lot of code. But low prio.) 

Also, it looks to me like nothing JetClass-specific is happening here anymore, i.e. the data is already in the same format as for our other tagging experiments. The only thing you do is to bring everything into the torch_geometric format. Can you confirm?

Reply:
Yes, that is essentially what I do. And I also could somehow move all of this to the dataset file. However, I cannot get rid of the for loop by developing a JCTaggingDataset class akin to the preeexistent ones, because I would need to load the whole dataset in memory. SimpleIterDataset by default loads a full batch at once, so we will have to adapt it so that the global token, beam, etc., are added without the for loop. It should be definitely possible, but it was non-trivial the first time I tried doing it.